frontend/src/liboperator/operator/operator.go

```diff
@@ -480,6 +485,94 @@ func openwrt(w http.ResponseWriter, r *http.Request, pool *DevicePool) {
 	proxy.ServeHTTP(w, r)
 }
 
+// WebSocket endpoint that proxies ADB
+func adbProxy(w http.ResponseWriter, r *http.Request, pool *DevicePool) {
+	vars := mux.Vars(r)
+	devId := vars["deviceId"]
+	dev := pool.GetDevice(devId)
+
+	if dev == nil {
+		http.Error(w, "Device not found", http.StatusNotFound)
+		return
+	}
+
+	devInfo := dev.privateData.(map[string]interface{})
+
+	// Find adb port for the device.
+	adbPort := 0
+	if adb_port, ok := devInfo["adb_port"]; ok {
+		adbPort = int(adb_port.(float64))
+	} else {
+		http.Error(w, "Cannot find adb port for the device", http.StatusNotFound)
+		return
+	}
+
+	// Prepare WebSocket and TCP socket for ADB
+	tcpConn, err := net.Dial("tcp", fmt.Sprintf("127.0.0.1:%v", adbPort))
+	if err != nil {
+		log.Print("Error while connect to ADB: ", err)
+		return
+	}
+	defer tcpConn.Close()
+
+	upgrader := websocket.Upgrader{}
+	wsConn, err := upgrader.Upgrade(w, r, nil)
+	if err != nil {
+		log.Print("Error while upgrading to WebSocket: ", err)
+		return
+	}
+	wsWrapper := &wsIoWrapper{
+		wsConn: wsConn,
+		pos:    0,
+		buf:    nil,
+	}
+	defer wsWrapper.Close()
+
+	// Redirect WebSocket to ADB tcp socket
+	go func() {
+		io.Copy(wsWrapper, tcpConn)
+		wsWrapper.Close()
```

Comment:
Should this be `tcpConn.Close()`? The `wsWrapper` is already closed on line 529.

Reply:
When I tested it, deferred wsWrapper.Close() was not called with spawned go routine, so it need to close explicitly io.Copy() call at line 536 to be exited when tcpConn is closed.